pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go

```diff
@@ -222,136 +306,220 @@ func (m *rangefeedMuxer) startNodeMuxRangeFeed(
-			// producer.muxClientCtx.Done()). That's fine -- we may have encountered range
-			// split or similar rangefeed error, causing the caller to exit (and
-			// terminate this stream), but the server side stream termination is async
-			// and probabilistic (rangefeed registration output loop may have a
-			// checkpoint event available, *and* it may have context cancellation, but
-			// which one executes is a coin flip) and so it is possible that we may
-			// see additional event(s) arriving for a stream that is no longer active.
-			if producer == nil {
-				if log.V(1) {
-					log.Infof(ctx, "received stray event stream %d: %v", e.StreamID, e)
+		switch t := event.GetValue().(type) {
+		case *kvpb.RangeFeedCheckpoint:
+			if t.Span.Contains(active.Span) {
+				// If we see the first non-empty checkpoint, we know we're done with the catchup scan.
+				if !t.ResolvedTS.IsEmpty() && active.catchupRes != nil {
+					active.catchupRes.Release()
+					active.catchupRes = nil
 				}
-				continue
+				// Note that this timestamp means that all rows in the span with
+				// writes at or before the timestamp have now been seen. The
+				// Timestamp field in the request is exclusive, meaning if we send
+				// the request with exactly the ResolveTS, we'll see only rows after
+				// that timestamp.
+				active.startAfter.Forward(t.ResolvedTS)
 			}
+		case *kvpb.RangeFeedSSTable:
+		case *kvpb.RangeFeedError:
+			log.VErrEventf(ctx, 2, "RangeFeedError: %s", t.Error.GoError())
+			if active.catchupRes != nil {
+				m.ds.metrics.RangefeedErrorCatchup.Inc(1)
+			}
+			conn.streams.Delete(event.StreamID)
+			if _, err := handleRangefeedError(ctx, t.Error.GoError()); err != nil {
+				// If this is an error we cannot recover from, terminate the rangefeed.
+				return err
+			}
+			if err := m.restartActiveRangeFeed(ctx, active, t.Error.GoError()); err != nil {
+				return err
+			}
+			continue
+		}
 
-			select {
-			case <-ctx.Done():
-				return ctx.Err()
-			case producer.eventCh <- &e.RangeFeedEvent:
-			case <-producer.muxClientCtx.Done():
-				if log.V(1) {
-					log.Infof(ctx, "received stray event stream %d: %v", e.StreamID, e)
+		active.onRangeEvent(nodeID, event.RangeID, &event.RangeFeedEvent)
+		msg := RangeFeedMessage{RangeFeedEvent: &event.RangeFeedEvent, RegisteredSpan: active.Span}
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case m.eventCh <- msg:
+		}
+
+		// Piggyback on this loop to check if any of the active ranges
+		// on this node appear to be stuck.
+		// NB: this does not notify the server in any way.  We may have to add
+		// a more complex protocol -- or better yet, figure out why ranges
+		// get stuck in the first place.
+		if timeutil.Now().Before(nextStuckCheck) {
+			if threshold := stuckThreshold(); threshold > 0 {
+				if err := conn.eachStream(func(id int64, a *activeMuxRangeFeed) error {
+					if !a.startAfter.IsEmpty() && timeutil.Since(a.startAfter.GoTime()) > stuckThreshold() {
```

Comment:
Won't this misfire if the `eventCh` send above is slow? Consider moving it higher up so that it primarily reacts to server events. Although I guess it doesn't matter all that much since we'll have head-of-line blocking between streams anyway which can cause false positives.